src/plugins/oidc/login.ts

```diff
@@ -57,70 +44,142 @@ const authorize = async (org: OrgData, scope: string) => {
  * authorization server will return "authorization_pending", in which case this
  * function will return undefined.
  */
-const fetchOidcToken = async (org: OrgData, authorize: AuthorizeResponse) => {
-  if (org.providerType === undefined) {
-    throw "Login requires a configured provider type.";
-  }
-  const init = {
-    method: "POST",
-    headers: OIDC_HEADERS,
-    body: urlEncode({
-      client_id: org.clientId,
-      device_code: authorize.device_code,
-      grant_type: DEVICE_GRANT_TYPE,
-    }),
-  };
-  validateProviderDomain(org);
-  const url =
-    org.providerType === "okta"
-      ? `https:${org.providerDomain}/oauth2/v1/token`
-      : org.providerType === "ping"
-        ? `https://${org.providerDomain}/${org.environmentId}/as/token`
-        : throwAssertNever(org.providerType);
+export const fetchOidcToken = async <T>(request: {
+  url: string;
+  init: RequestInit;
+}) => {
+  const { url, init } = request;
   const response = await fetch(url, init);
-
   if (!response.ok) {
     if (response.status === 400) {
       const data = await response.json();
       if (data.error === "authorization_pending") return undefined;
     }
     await validateResponse(response);
   }
-  return (await response.json()) as TokenResponse;
+  return (await response.json()) as T;
+};
+
+const providerType: (org: OrgData) => NonNullable<string> = (org) => {
```

Comment:
```suggestion
const providerType: (org: OrgData) => string = (org) => {
```

`string` is already non-nullable